core/src/main/java/org/apache/calcite/schema/Schema.java

```diff
@@ -56,9 +59,30 @@
  * {@link Schema#getSubSchema(String)}.
  */
 public interface Schema {
+
+  /**
+   * Returns a lookup object to find tables.
+   *
+   * @return Lookup
+   */
+  default Lookup<Table> tables() {
+    return new CompatibilityLookup<>(this::getTable, this::getTableNames);
+  }
+
+  /**
+   * Returns a lookup object to find sub schemas.
+   *
+   * @return Lookup
+   */
+  default Lookup<? extends Schema> subSchemas() {
+    return new CompatibilityLookup<>(this::getSubSchema, this::getSubSchemaNames);
+  }
+
   /**
    * Returns a table with a given name, or null if not found.
    *
+   * <p>Please use {@link Schema#tables()} and {@link Lookup#get(String)} instead.
```

Comment:
people will not find this comment.
maybe the API should be marked as deprecated.
on the other hand, this API makes sense for some cases - and it should perhaps have a default implementation doing exactly what you suggest.
There must be some downside to doing that - what is it?

Reply:
If there is consensus that the method should be deprecated I see no problem in marking it as such, we do it on a regular basis already. Given the pervasive use of Calcite in many downstream projects, we haven't removed any deprecated method so far (there are several comments saying "to be removed before v2.0 in the codebase"), but the fact that downstream projects are warned ahead of time is exactly what we want.

Reply:
For me, it would be OK to add the `@Deprecated` annotation. 

Since it's my first PR that introduces an incompatible change, I wanted to wait for others' opinions first.

Reply:
If these methods are not used anywhere anymore, then it may be good to mark them as deprecated, and document this in history.md.